### PR TITLE
[proxima-core] #142 introduce operator specific transformations

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

File: core/src/main/java/cz/o2/proxima/transform/ContextualTransformation.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+package cz.o2.proxima.transform;
+
+import cz.o2.proxima.annotations.Evolving;
+import cz.o2.proxima.repository.DataOperator;
+import cz.o2.proxima.repository.Repository;
+import java.util.Map;
+
+@Evolving
+public interface ContextualTransformation<OP extends DataOperator>
```
Comment (LesTR) [inline core/src/main/java/cz/o2/proxima/transform/ContextualTransformation.java]: Difference between ContextualTransform and ContextualTransformation? Maybe better naming?

File: beam/core/src/test/java/cz/o2/proxima/beam/core/JoinProxyTransform.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+public class JoinProxyTransform extends BeamProxyTransform {
```
Comment (frozen22) [inline beam/core/src/test/java/cz/o2/proxima/beam/core/JoinProxyTransform.java]: Missing Javadoc for public class :)

File: core/src/main/java/cz/o2/proxima/transform/DelegatedTransform.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import cz.o2.proxima.annotations.Internal;
+import cz.o2.proxima.repository.DataOperator;
+import cz.o2.proxima.repository.DataOperatorFactory;
+import java.io.Serializable;
+
+@Internal
+public interface DelegatedTransform extends Serializable {
+
+  /** Verify if this {@link DataOperator} specific (contextual) transform. */
+  default boolean isContextualTransform() {
```
Comment (frozen22) [inline core/src/main/java/cz/o2/proxima/transform/DelegatedTransform.java]: Hmm, `isContextualTransform` in `DelegatedTransform`, typo?

Comment (frozen22) [reply core/src/main/java/cz/o2/proxima/transform/DelegatedTransform.java]: I'm seeing it now, this is parent of ContextualTransform...

File: beam/core/src/main/java/cz/o2/proxima/beam/core/AttributeFamilyProxyDataAccessor.java [hunk 4]
```diff
@@ -139,4 +159,32 @@ private StreamElement transformSingleRead(StreamElement input) {
...
+    if (proxies.stream().allMatch(p -> p.getReadTransform().isContextualTransform())) {
+      List<ProxyTransform> transforms =
+          proxies
+              .stream()
+              .map(AttributeProxyDescriptor::getReadTransform)
+              .distinct()
+              .collect(Collectors.toList());
+      Preconditions.checkArgument(
+          transforms.size() == 1,
+          "When reading using BeamProxyTransform only single attribute on input is allowed, got [%s]",
```
[the remark this answers is not in the document]
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/core/AttributeFamilyProxyDataAccessor.java]: :+1:

File: beam/core/src/main/java/cz/o2/proxima/beam/core/AttributeFamilyProxyDataAccessor.java [hunk 5]
```diff
@@ -139,4 +159,32 @@ private StreamElement transformSingleRead(StreamElement input) {
...
+          proxies);
+
+      Preconditions.checkArgument(
+          transforms.get(0) instanceof BeamProxyTransform,
+          "Please do not mix multiple contextual proxies in single config, expected class [%s] got [%s]",
```
[the remark this answers is not in the document]
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/core/AttributeFamilyProxyDataAccessor.java]: Removed `Please`. There's no need to be polite here! :-)

File: beam/core/src/main/java/cz/o2/proxima/beam/core/BeamProxyTransform.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+package cz.o2.proxima.beam.core;
+
+import cz.o2.proxima.repository.DataOperatorFactory;
+import cz.o2.proxima.storage.StreamElement;
+import cz.o2.proxima.storage.commitlog.Position;
+import cz.o2.proxima.transform.ContextualProxyTransform;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.values.PCollection;
+
+public abstract class BeamProxyTransform implements ContextualProxyTransform<BeamDataOperator> {
```
[the remark this answers is not in the document]
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/core/BeamProxyTransform.java]: :+1:

File: beam/core/src/test/resources/test-beam-proxy.conf [hunk 7]
```diff
@@ -0,0 +1,13 @@
+entities {
+  gateway {
+    attributes {
+      armed-proxy {
+        // read gateway.armed, transform to length of the payload,
+        // but only when gateway.state is not null
```
[the remark this answers is not in the document]
Comment (je-ik) [reply beam/core/src/test/resources/test-beam-proxy.conf]: Yes, `gateway.status`, fixed. `scheme` is the same as for any other attribute.

File: core/src/main/java/cz/o2/proxima/repository/AttributeDescriptor.java [hunk 8]
```diff
@@ -86,10 +88,11 @@ static Builder newBuilder(Repository repo) {
       ProxyTransform transformRead,
       AttributeDescriptor<T> targetWrite,
       ProxyTransform transformWrite,
-      boolean replica) {
+      boolean replica,
+      ValueSerializer<T> valueSerializer) {
```
[the remark this answers is not in the document]
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/repository/AttributeDescriptor.java]: It is defined by `scheme` of the proxy attribute. Proxy attribute can have different scheme than attribute it is constructed from.

File: core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java [hunk 9]
```diff
@@ -329,6 +315,8 @@ public final void reloadConfig(boolean loadFamilies, Config conf) {
       loadProxiedFamilies(true);
       // Read transformations from one entity to another.
       readTransformations(config);
+      // call setuo() on all transforms
```
[the remark this answers is not in the document]
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java]: :+1:

File: beam/core/src/test/java/cz/o2/proxima/beam/core/JoinProxyTransform.java (hunk 2 above)
Comment (je-ik) [reply beam/core/src/test/java/cz/o2/proxima/beam/core/JoinProxyTransform.java]: That's inside tests I think we don't have to enforce too strict rules. :)

File: core/src/main/java/cz/o2/proxima/transform/DelegatedTransform.java (hunk 3 above)
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/transform/DelegatedTransform.java]: I changed this to be more readable.

**Comment by je-ik** [comment]
@LesTR @frozen22 pushed review

**Comment by je-ik** [comment]
@frozen22 I cannot re-request review from you (don't know why). Can you please have a look nevertheless?

File: beam/core/src/test/java/cz/o2/proxima/beam/core/JoinProxyTransform.java (hunk 2 above)
Comment (frozen22) [reply beam/core/src/test/java/cz/o2/proxima/beam/core/JoinProxyTransform.java]: ok:)

File: core/src/main/java/cz/o2/proxima/transform/DelegatedTransform.java (hunk 3 above)
Comment (frozen22) [reply core/src/main/java/cz/o2/proxima/transform/DelegatedTransform.java]: Thx

**Comment by je-ik** [comment]
This closes O2-Czech-Republic#142

**Pull request merged**